File: cplus/leetcode/202501/algo3003.cpp
#include <bits/stdc++.h>
using namespace std;

// 周赛 379
class Solution {
public:
    int maxPartitionsAfterOperations(string s, int k) {
        if (k == 26) return 1;
        unordered_map<long long, int> memo;

        function<int(int, int, bool)> dfs = [&](int i, int mask, bool changed) -> int {
            if (i == s.size()) return 1;

            long long argMask = (long long) i << 32 | mask << 1 | changed;
            // 两种写法
            // if (memo.count(argMask)) return memo[argMask];
            auto it = memo.find(argMask);
            if (it != memo.end()) return it->second;
            
            int res = 0;
            // don't change i 
            if (__builtin_popcount(mask | 1 << (s[i]-'a')) > k) 
                res = dfs(i+1, 1 << (s[i]-'a'), changed) + 1;
            else res = dfs(i+1, mask | 1 << (s[i]-'a'), changed);

            // change i
            if (!changed) {
                for (int ch=0; ch<26; ch++) {
                    if (__builtin_popcount(mask | 1 << ch) > k) 
                        res = max(res, dfs(i+1, 1 << ch, true) + 1);
                    else res = max(res, dfs(i+1, mask | 1 << ch, true));
                }
            }
            
            return memo[argMask] = res;
        };

        return dfs(0,0,false);    

    }
};

File: python/algo/202501/algo3003.py
from functools import cache


class Solution:
    def maxPartitionsAfterOperations(self, s: str, k: int) -> int:
        @cache
        def dfs(i:int, mask:int, changed:bool) -> int:
            if i == len(s): return 1

            bit = ord(s[i]) - ord('a')

            # not change
            res = 0
            if (mask | 1 << bit).bit_count() > k:
                res = dfs(i+1, 1<<bit, changed) + 1
            else: res = dfs(i+1, mask | 1 << bit, changed)

            # change it
            if changed: return res
            for j in range(26):
                if j == bit: continue
                # 犯了弱智错误， "=" 写成 "=="
                # newmask == mask | 1 << j 引以为戒
                if (mask | 1 << j).bit_count() > k:
                    res = max(res, dfs(i+1, 1 << j, True) + 1)
                else: res = max(res, dfs(i+1, mask | 1 << j, True))
            return res
        
        return dfs(0, 0, False)